mitodl/unified-ecommerce
# Adds transaction/purchasing workflow and support for API gateway

### What are the relevant tickets?

https://github.com/mitodl/hq/issues/3166
https://github.com/mitodl/hq/issues/3158

### Description (What does it do?)

Data model:
- Adds Basket and Order models, and their associated proxy models. These are lifted out of MITx Online.
- Updates IntegratedSystem to include a slug field.
- Adds/update management commands for products and systems appropriately.

API Gateway/Authentication:
- Includes a middleware that can pull auth from `X-UserInfo` header if that exists
- Includes a middleware that can pull auth from `X-ForwardedUser` header if that exists
- Adds some tasks to pull user data via Keycloak for a given user
- Adds infrastructure pieces to optionally run APISIX and/or Traefik with the app

Basket and Orders:
- Adds a cart test mule app
- Completes the purchasing workflow; you should be able to add products to the basket and check out successfully

Meta:
- Updates the README
- Added a viewset that allows for different result sets based on user and session
- Tests for most everything (not cart test mule app)


### Screenshots (if appropriate):

![Screenshot 2024-02-15 at 9 38 57 AM](https://github.com/mitodl/unified-ecommerce/assets/945611/266b9443-811d-431e-a0f8-d6584dd44cd2)

### How can this be tested?

Automated tests should work.

To manually test, you will need a working Keycloak instance (see the README for info), and you need to set up Traefik (preferred). (This can also work with APISIX but as noted in the README the instance included won't have any routes, so Traefik is easier to get working.) You will also need CyberSource keys (which can be pulled from another app) and you'll need to create a system and products. You should then be able to access the cart test mule at `/cart`. This should work as expected - you should be able to add to cart and check out and that all should work as expected. There are no receipt emails, post-sale events, or refunding at this point. 

You should be able to test with the Swagger UI as well.

Model auth is still in place so you can log in with a local user if you need to. The app should still be listening on port 8073; going there should not bounce you into Keycloak.

SSO users will be created without staff or superuser, but you can change that if you wish. The usernames will be the UUIDs created by Keycloak once the app has retrieved information from it.

### Additional Context

The cart test mule app is just a simple app to demonstrate the cart workflow - it isn't done in a proper fashion for OL apps. The app that it's in will become the user frontend for the Unified Ecommerce system, but that'll be for a future epic. But, given its status now, there's no tests for it and it's pretty limited in functionality.
